### PR TITLE
ENH/REF: Markov switching in log space: Hamilton filter / Kim smoother

### LINKED ISSUE #4125: ENH: markov_switching switching needs to be in log scale
The filters and smoothers in statsmodels.tsa.regime_switching.markov_switching (e.g., py_hamilton_filter) should be in logscale and use (scipy.misc.logsumexp etc).

The current filters (e.g. Lines 187--201) are all done in linear scale.  This can results in NaNs from underflow when a long sequence of observations with obvious states are encountered.  The standard robust implementation of HMMs is done in logscale.  See for example:
http://bozeman.genome.washington.edu/compbio/mbt599_2006/hmm_scaling_revised.pdf

### LINKED ISSUE #5380: BUG: MarkovRegression failed to converge (underflow error)
Hi!
I am trying to fit MarkovRegression model to (financial) data and for some of the assets it fails to converge. 
My code:
`m = sm.tsa.MarkovRegression(series, k_regimes=2, switching_variance=True).fit(disp=True)`

 Output:
```
Warning: Desired error not necessarily achieved due to precision loss.
         Current function value: nan
         Iterations: 0
         Function evaluations: 1
         Gradient evaluations: 1
C:\Program Files\Anaconda\lib\site-packages\statsmodels\tsa\regime_switching\markov_switching.py:391: RuntimeWarning: invalid value encountered in true_divide
  predicted_joint_probabilities[..., t+1])
C:\Program Files\Anaconda\lib\site-packages\numpy\linalg\linalg.py:1876: RuntimeWarning: invalid value encountered in greater
  large = s > cutoff
C:\Program Files\Anaconda\lib\site-packages\numpy\core\fromnumeric.py:83: RuntimeWarning: invalid value encountered in reduce
  return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
C:\Program Files\Anaconda\lib\site-packages\statsmodels\tools\numdiff.py:96: RuntimeWarning: invalid value encountered in maximum
  h = EPS**(1. / s) * np.maximum(np.abs(x), 0.1)
C:\Program Files\Anaconda\lib\site-packages\scipy\special\_logsumexp.py:118: RuntimeWarning: invalid value encountered in log
  out = np.log(s)
C:\Program Files\Anaconda\lib\site-packages\statsmodels\tsa\regime_switching\markov_regression.py:190: RuntimeWarning: invalid value encountered in true_divide
  np.exp(-0.5 * resid**2 / variance) / np.sqrt(2 * np.pi * variance))
C:\Program Files\Anaconda\lib\site-packages\statsmodels\tsa\regime_switching\markov_switching.py:1053: RuntimeWarning: invalid value encountered in log
  return np.log(results[5])
C:\Program Files\Anaconda\lib\site-packages\statsmodels\base\model.py:496: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  "Check mle_retvals", ConvergenceWarning)
```

This is the case for around 70 time series out of 850 in my dataset. I tried other methods, e.g. 'powell', but with no success. It looks like `predicted_joint_probabilities[..., t+1]``gives `nan` or `inf` values. There are no outliers in the data.

### PR DESCRIPTION
**Pull request by ChadFulton**
- [x] closes #4125
- [x] closes #5380
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message.

This rewrites the Hamilton filter and Kim smoother in log-space to avoid underflow errors in Markov switching models.

Adds a unit test based on the report in #5380.

Note: in one commit I just added the new filter / smoother, and then in the next commit I removed all of the historical versions (there had been several sort of piling up in there). This way we'll have all the historical versions in git, but the files in master will be a little more lightweight.

I'm thinking about waiting for v0.11, but could be convinced to include it in v0.10. On the one hand, it's a helpful bug fix, and so maybe it should go in sooner than later. On the other hand, it won't have been used much by the time v0.10 is released (presumably).

### CODE REVIEW

**Comment by ChadFulton** [comment]
Hmm. somehow messed up the git log.

**Comment by ChadFulton** [comment]
I should add that this implies a performance hit. Latest commit releases the GIL in the inner loop, which helps a bit. However, old code was about twice as fast.

Still, probably worth it to avoid the underflow problem which seems to pop up relatively often.

File: statsmodels/tsa/regime_switching/_hamilton_filter.pyx.in [hunk 1]
```diff
@@ -24,72 +24,107 @@ import numpy as np
...
-                                  joint_likelihoods,
-                                  predicted_joint_probabilities[:, t],
-                                  filtered_joint_probabilities[:, t],
-                                  filtered_joint_probabilities[:, t+1])
-
-
-cdef {{prefix}}hamilton_filter_iteration(int t, int k_regimes, int order,
+    tmp_predicted_joint_probabilities = np.zeros(k_regimes, dtype={{dtype}})
+
+    # zlog, zexp require the GIL (at least for now), so we can only release
```
Comment (bashtage) [inline statsmodels/tsa/regime_switching/_hamilton_filter.pyx.in]: You should be able to add nogil to the end of the functions in .src.math and then you can simplify this to always use nogil.

Comment (ChadFulton) [reply statsmodels/tsa/regime_switching/_hamilton_filter.pyx.in]: Ah nice, thanks!

**Comment by ChadFulton** [comment]
Rebased to hopefully get rid of the unrelated linting errors.

I plan to merge this once the release has been fully processed and master is open for PRs again.

**Pull request merged**